### DISCUSSION TITLE
How to implement SearchQueryInterrupter

### DISCUSSION THREAD

**Discussion by kyowanghwe** [dentsusoken/iPLAss#1655] [Q&A]
For example, my search query return a list of employee id=1,2,3,4 and i just want to return only id=1,2 and hide id=3,4
So how can i do

**Comment by KoheiKGYM** [comment]
Hi @kyowanghwe,
thanks for using iPLAss!

Using beforeSearch in SearchQueryInterrupter allows you to add search conditions before executing the search, enabling you to narrow down the search results. On the other hand, afterSearch in SearchQueryInterrupter is called for each record after the search, so it is not possible to filter the items from the search results.
https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#GemCustomize_Search:~:text=%E5%87%A6%E7%90%86%E3%81%AE%E7%B5%84%E3%81%BF%E8%BE%BC%E3%81%BF-,5.1.%20%E3%82%AB%E3%82%B9%E3%82%BF%E3%83%A0%E6%A4%9C%E7%B4%A2%E5%87%A6%E7%90%86,-%E6%A4%9C%E7%B4%A2%E4%B8%80%E8%A6%A7%E3%82%84
(Please note that the documentation is currentlly available only in Japanese. I apologize for the inconvenience, but please use a translation tool for the time being.)

A class that implements SearchQueryInterrupter should be specified in **Admin Console > Search Layout > Search Screen > EQL Custom Processing Class Name**.
https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#GemCustomize_Search:~:text=%E3%81%95%E3%82%8C%E3%81%BE%E3%81%99%E3%80%82-,EQL%E3%82%AB%E3%82%B9%E3%82%BF%E3%83%A0%E5%87%A6%E7%90%86%E3%82%AF%E3%83%A9%E3%82%B9%E5%90%8D,-%E6%A4%9C%E7%B4%A2%E6%99%82%E3%82%84
![image](https://github.com/user-attachments/assets/e8aec700-225d-424c-83bc-d43b902c66cb)

Here is a sample of filtering using beforeSearch:
```java
package org.iplass.mtp;

import org.iplass.mtp.view.generic.SearchQueryInterrupter;
...

public class Test implements SearchQueryInterrupter {

@Override
public SearchQueryContext beforeSearch(RequestContext request, SearchFormView view, Query query, SearchQueryType type) {
  Condition condition = query.getWhere().getCondition();

  And and = new And();
  if (condition != null) {
   and.and(condition);
  }

  List<Integer> employeeIds = Arrays.asList(1, 2);
  Condition inCondition = new In("employee_id", employeeIds.toArray());
  and.and(inCondition);

  Where where = new Where();
  where.setCondition(and);
  query.setWhere(where);

  SearchQueryContext context = new SearchQueryContext(query);

  return context;
 }

}
```

If the requirement is simply to add search conditions, you can use Search Default Condition without SearchQueryInterrupter.
By specifying conditions in **Search Layout > Search Condition > Search Default Condition**, you can easily apply filtering conditions before executing the search.
https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#default_search_cond:~:text=%E3%83%81%E3%82%A7%E3%83%83%E3%82%AF%E3%81%97%E3%81%BE%E3%81%99%E3%80%82-,%E3%83%87%E3%83%95%E3%82%A9%E3%83%AB%E3%83%88%E6%A4%9C%E7%B4%A2%E6%9D%A1%E4%BB%B6,-PreparedQuery%E5%BD%A2%E5%BC%8F%E3%81%A7
![image](https://github.com/user-attachments/assets/453fc03e-f005-4ecc-8b78-11c1f24ce428)
